### PR TITLE
Fix #632: Add --no-stream to ipfs-cluster-ctl

### LINKED ISSUE #632: Proper JSON output from ipfs-cluster-ctl
* [x] Version information (mark as appropiate):
  * [x] Master
  * [x] Release candidate for next version
  * [x] Latest stable version
  * [ ] An older version I should not be using
* [x] Type (mark as appropiate):
  * [ ] Bug
  * [x] Feature request
  * [ ] Enhancement
* [x] Operating system (mark as appropiate):
  * [x] Linux
  * [x] macOS
  * [x] Windows
  * [ ] Other: `which?`
* [x] Installation method (mark as appropiate):
  * [x] Binaries from dist.ipfs.io
  * [x] Built from sources
  * [x] Docker
  * [x] Snap
  * [ ] Other: `which?`

#### Description
I am programmatically adding the [benchmark results](https://benchmarks.ipfs.team) with [`ipfs-cluster-ctl`](https://github.com/ipfs/benchmarks/blob/master/runner/ipfs.js#L15) and need to return the SHA for presentation purposes. The current output with `--enc json` needs [additional work](https://github.com/ipfs/benchmarks/blob/master/runner/ipfs.js#L27) before it can be parsed as JSON. 

```
{"code":0,"message":"","Name":"test1.jpg","Hash":"QmPxBet4MHo93b9nCE2FYRtDTowVA3bMRdDBZcYAiBkZKN","Size":"4696045"}
{"code":0,"message":"","Name":"test2.jpg","Hash":"QmfJ8fzLwCmuvxNG8RH41RmmjGzMrXw38pjP85EJSiXKJ4","Size":"4696045"}
```
_Note the missing surrounding `[]` and commas between the objects._

As per [this issue](https://github.com/ipfs/ipfs-cluster/issues/526#issuecomment-418120772) I'd like to request a flag to return `non-streaming` json output which can be parsed by `JSON.parse` as is.

> Hi, this is not a bug (it's a feature). /add response is a streaming response and mimics what ipfs /add does (for the moment). Perhaps you'd like to turn this into a feature request asking to support a stream=false parameter and return a single complete json response, or something else, whatever your usecase is, and we'll consider it if there is good reasoning behind it.

### PR DESCRIPTION
**Pull request by hsanjuan**
tl;dr: this solves the user's immediate need and, even if not tne strictest
solution, it is the simplest one.

I think we should not have the server buffer output when we can do it rather
easily client side and have the clients use their own memory for the task even
if `stream-channels=false` would do this.

We can always change the approach, but this is the minimal solution to
json array with all the AddedOutput things.

We might not buffer at all and hack a `[`, `,` separating elements and `]`
at the end, when json encoding is enabled, etc. But that would not be clean,
specially if we mean to support more output formats at some point.

Enabling supporting stream-channels=false in the api/rest/client means adding
new methods, with tests, modifying the interface etc etc. for what is
essentially a presentation issue in "ctl" in the end. Similarly we could
buffer inside the client, but it is so trivial that I don't see advatange.
We should also start thinking about moving to streaming API endpoints and
when that moment arrives we shall revisit this discussion.

I have removed the hacky manual output parts by declaring a custom
addedOutputQuiet type which wraps added output and is understood by
the formatters.go helpers.

License: MIT
Signed-off-by: Hector Sanjuan <code@hector.link>

### CODE REVIEW

File: cmd/ipfs-cluster-ctl/formatters.go [hunk 1]
```diff
@@ -25,6 +30,10 @@ func jsonFormatObject(resp interface{}) {
 		jsonFormatPrint(resp.(api.Pin).ToSerial())
 	case api.AddedOutput:
 		jsonFormatPrint(resp.(api.AddedOutput))
+	case addedOutputQuiet:
+		// print original object as in json it does
```
Comment (lanzafame) [inline cmd/ipfs-cluster-ctl/formatters.go]: nitpick: comment needs rewording

File: cmd/ipfs-cluster-ctl/main.go [hunk 2]
```diff
@@ -277,6 +277,10 @@ If you prefer faster adding, add directly to the local IPFS and trigger a
 					Name:  "quieter, Q",
 					Usage: "Write only final hash to output",
 				},
+				cli.BoolFlag{
+					Name:  "no-stream",
+					Usage: "Buffer output locally. Produces a valid slice with --enc=json.",
```
Comment (lanzafame) [inline cmd/ipfs-cluster-ctl/main.go]: nitpick: s/valid slice/valid JSON slice

File: cmd/ipfs-cluster-ctl/main.go [hunk 3]
```diff
@@ -391,29 +395,36 @@ If you prefer faster adding, add directly to the local IPFS and trigger a
 				wg.Add(1)
 				go func() {
 					defer wg.Done()
-					var last *api.AddedOutput
+
+					var buffered []addedOutputQuiet
+					var lastBuf = make([]addedOutputQuiet, 1, 1)
+					var qq = c.Bool("quieter")
+					var q = c.Bool("quiet") || qq
```
Comment (kishansagathiya) [inline cmd/ipfs-cluster-ctl/main.go]: On running with `quiet` and `quieter`, it is also printing name and size.

File: cmd/ipfs-cluster-ctl/main.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (hsanjuan) [reply cmd/ipfs-cluster-ctl/main.go]: hmm good catch. I changed the name of the flag and introduced a bug. The `--no-stream` output should produce a `[ .. ]` result printed at the end. The streaming output prints AddedOuput object as they arrive.

I want to name it `no-stream` because it relates to how results are streamed to the user. And might eventually mean that this will use no-streaming endpoint from the server in the future.

File: cmd/ipfs-cluster-ctl/main.go [hunk 4]
```diff
@@ -391,29 +395,36 @@ If you prefer faster adding, add directly to the local IPFS and trigger a
...
+					var bufferResults = !c.Bool("stream")
```
[the remark this answers is not in the document]
Comment (hsanjuan) [reply cmd/ipfs-cluster-ctl/main.go]: yes

File: cmd/ipfs-cluster-ctl/main.go (hunk 2 above)
Comment (hsanjuan) [reply cmd/ipfs-cluster-ctl/main.go]: ok

File: cmd/ipfs-cluster-ctl/formatters.go (hunk 1 above)
Comment (hsanjuan) [reply cmd/ipfs-cluster-ctl/formatters.go]: done

File: cmd/ipfs-cluster-ctl/main.go (hunk 3 above)
Comment (hsanjuan) [reply cmd/ipfs-cluster-ctl/main.go]: should be fixed to after `--no-stream`

File: cmd/ipfs-cluster-ctl/main.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (hsanjuan) [reply cmd/ipfs-cluster-ctl/main.go]: > why not always buffer locally?

In the general case, you are just using memory on stuff that you can flush out. Mid-term plan will be to move any endpoints listing things to streaming, as otherwise it's unpractical to deal with huge pinsets.

File: cmd/ipfs-cluster-ctl/main.go [hunk 5]
```diff
@@ -277,6 +277,10 @@ If you prefer faster adding, add directly to the local IPFS and trigger a
 					Name:  "quieter, Q",
 					Usage: "Write only final hash to output",
```
Comment (kishansagathiya) [inline cmd/ipfs-cluster-ctl/main.go]: Nothing to do with this PR, but this description is wrong. `quieter` doesn't print just hash. It also prints name and size(that's the behaviour I see in master). It's a little that I think we can push along with this PR.

Comment (hsanjuan) [reply cmd/ipfs-cluster-ctl/main.go]: It does just print the hash? Unless you mean in the json output, to which this does not really apply because we print the json objects.

```
[18:30:23] ~/ipfstest $ ipfs-cluster-ctl  add -Q -r testsh/
QmekoCaPuQ8wcWhuws4KRZnNhzaTNrbM1e6dqgWRVLyZnw
[18:30:28] ~/ipfstest $ ipfs-cluster-ctl  add --no-stream -Q -r testsh/
QmekoCaPuQ8wcWhuws4KRZnNhzaTNrbM1e6dqgWRVLyZnw
```

Comment (kishansagathiya) [reply cmd/ipfs-cluster-ctl/main.go]: yup, I was talking about the json output

Comment (hsanjuan) [reply cmd/ipfs-cluster-ctl/main.go]: yeah, json output should print valid json in any case. -q and -Q have text-formatting implications which do not apply to json (here or in `ipfs add`)

**Comment by hsanjuan** [comment]
Not releasing today so this can wait for your OK @lanzafame
